analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/RestrictDeserializedTypesBase.cs

```diff
@@ -26,4 +28,167 @@ public abstract class RestrictDeserializedTypesBase : SymbolicRuleCheck
+    private static readonly KnownType[] TypesWithDeserializeMethod = FormattersWithBinder.Concat(new[] { JavaScriptSerializer }).ToArray();
+
+    private readonly Dictionary<ISymbol, SyntaxNode> additionalLocationsForSymbols = new();
+    private readonly Dictionary<IOperation, SyntaxNode> additionalLocationsForOperations = new();
+
+    protected override ProgramState PostProcessSimple(SymbolicContext context)
+    {
+        var state = context.State;
+        var operation = context.Operation.Instance;
+        if (operation.Kind == OperationKindEx.ObjectCreation)
+        {
+            return operation.Type.IsAny(FormattersWithBinder)
+                ? state.SetOperationConstraint(operation, SerializationConstraint.Unsafe)
+                : ProcessOtherSerializerCreations(state, operation.ToObjectCreation());
+        }
+        else if (operation.AsAssignment() is { } assignment)
+        {
+            if (ProcessBinderAssignment(state, assignment) is { } binderProcessedState)
+            {
+                return binderProcessedState;
+            }
+            else if (AdditionalLocation(state, assignment.Value) is { } methodDeclaration
+                && assignment.Target.TrackedSymbol() is { } symbol)
+            {
+                additionalLocationsForSymbols[symbol] = methodDeclaration;
+            }
+        }
+        else if (UnsafeDeserialization(state, operation) is { } invocation)
+        {
+            var methodDeclaration = AdditionalLocation(state, invocation.Instance);
+            var additionalLocations = methodDeclaration is not null
+                ? new[] { GetIdentifier(methodDeclaration).GetLocation() }
+                : Array.Empty<Location>();
+            ReportIssue(operation, additionalLocations, RestrictTypesMessage);
+        }
+        return state;
+    }
+
+    private ProgramState ProcessOtherSerializerCreations(ProgramState state, IObjectCreationOperationWrapper objectCreation)
+    {
+        if (UnsafeJavaScriptSerializer(state, objectCreation, out var resolveTypeDeclaration))
+        {
+            additionalLocationsForOperations[objectCreation.WrappedOperation] = resolveTypeDeclaration;
+            return state.SetOperationConstraint(objectCreation.WrappedOperation, SerializationConstraint.Unsafe);
+        }
+        else if (UnsafeLosFormatter(state, objectCreation))
+        {
+            ReportIssue(objectCreation.WrappedOperation, VerifyMacMessage);
+        }
+        return state;
+    }
+
+    private bool UnsafeJavaScriptSerializer(ProgramState state, IObjectCreationOperationWrapper objectCreation, out SyntaxNode resolveTypeDeclaration)
+    {
+        resolveTypeDeclaration = null;
+        if (objectCreation.Type.Is(JavaScriptSerializer))
+        {
+            foreach (var argument in objectCreation.Arguments.Select(x => state.ResolveCaptureAndUnwrapConversion(x.ToArgument().Value)))
```

Comment:
This can be simplified and done without a loop. The constructor can have either 0 or 1 argument (and it's a legacy class, so it's not likely to change).

```cs
private bool UnsafeJavaScriptSerializer(ProgramState state, IObjectCreationOperationWrapper objectCreation, out SyntaxNode resolveTypeDeclaration)
{
    resolveTypeDeclaration = null;
    if (objectCreation.Type.Is(JavaScriptSerializer)
        && objectCreation.Arguments.Length == 1
        && state.ResolveCaptureAndUnwrapConversion(objectCreation.Arguments[0].ToArgument().Value) is { } argument)
    {
            if (argument.Type.Is(KnownType.System_Web_Script_Serialization_SimpleTypeResolver))
            {
                return true;
            }
            else if (DeclarationCandidates(argument)?.FirstOrDefault(IsResolveType) is { } declaration
                && !ThrowsOrReturnsNull(declaration))
            {
                resolveTypeDeclaration = declaration;
                return true;
            }
    }
    return false;
}
```